torch/distributed/fsdp/flat_param.py

```diff
@@ -1017,18 +1147,297 @@ def unflatten_as_params(self) -> Generator:
+                    0,
+                    dtype=param.dtype,
+                    device=self.flat_param.device,
+                    requires_grad=False,
+                )
+        assert self.flat_param._shared_params is not None
+        for i, (
+            param,
+            (param_name, module, _, prim_param_name, prim_module, _),
+        ) in enumerate(
+            zip(self.flat_param._shared_params, self.flat_param._shared_param_infos)
+        ):
+            setattr(module, param_name, param)
+            prim_param = getattr(prim_module, prim_param_name)
+            param.data = prim_param  # could be both empty and non-empty
+
+    @torch.no_grad()
+    def _use_sharded_grad_views(self) -> None:
+        """
+        Sets the original module parameter variables' gradients to be flattened
+        views into the sharded flattened parameter's gradient. This is a no-op
+        if there is no gradient.
+
+        Parameters whose data is not present in the sharded flattened parameter
+        and parameters with ``requires_grad=False`` have their gradients set to
+        ``None``. Since the gradient variables do not need to be preserved,
+        this method does not manipulate existing ``Tensor`` data directly and
+        creates new ``Tensor`` variables instead.
+        """
+        flat_param = self.flat_param
+        self._check_sharded(flat_param)
+        # Priority: `_cpu_grad` > `_saved_grad_shard` > `grad`
+        # - CPU offloading: `_cpu_grad`
+        # - No CPU offloading + sharded strategies: `_saved_grad_shard`
+        # - No CPU offloading + `NO_SHARD`: `grad`
+        if hasattr(flat_param, "_cpu_grad"):
+            grad = flat_param._cpu_grad  # type: ignore[attr-defined]
+        elif hasattr(flat_param, "_saved_grad_shard"):
+            grad = flat_param._saved_grad_shard  # type: ignore[attr-defined]
+        else:
+            grad = flat_param.grad
+        if grad is None:
+            return  # no-op
+        self._check_sharded(grad)
+        start, end = self.flat_param._shard_indices  # type: ignore[attr-defined]
+        offset = 0
+        assert self.flat_param._params is not None
+        for i, param in enumerate(self.flat_param._params):
+            in_sharded_flat_param = (
+                i >= start
+                and i <= end
+                and self.flat_param._shard_param_offsets  # type: ignore[attr-defined]
+            )
+            if in_sharded_flat_param:
+                param_start, param_end = self.flat_param._shard_param_offsets[i - start]  # type: ignore[attr-defined]
+                numel_in_shard = param_end - param_start + 1
+                if param.requires_grad:
+                    param.grad = grad[offset : offset + numel_in_shard].reshape(
+                        param.shape
+                    )
+                else:
+                    param.grad = None
+                offset += numel_in_shard
+            else:
+                param.grad = None
```

Comment:
a lot of the actual setting grad as view code looks to be similar to setting the params, aside some things like grad = None if it does not exist in the shard versus parameter being an empty tensor. could we consolidate as much as possible?

Reply:
Yes, I agree that there is a lot of boilerplate looping over the parameters/gradients and checking if in the local shard.

If you do not mind, I may try to refactor that in a follow-up PR because it is actually not _exactly_ obvious how to coalesce them into one skeleton.